src/v/rpc/transport.cc

```diff
@@ -149,69 +158,98 @@ ss::future<> transport::connect() {
 
 ss::future<result<std::unique_ptr<streaming_context>>>
 transport::send(netbuf b, rpc::client_opts opts) {
+    return do_send(_seq++, std::move(b), std::move(opts));
+}
+
+ss::future<result<std::unique_ptr<streaming_context>>>
+transport::do_send(uint32_t seq, netbuf b, rpc::client_opts opts) {
     using ret_t = result<std::unique_ptr<streaming_context>>;
     // hold invariant of always having a valid connection _and_ a working
     // dispatch gate where we can wait for async futures
     if (!is_valid() || _dispatch_gate.is_closed()) {
+        _last_seq = std::max(_last_seq, seq);
         return ss::make_ready_future<ret_t>(errc::disconnected_endpoint);
     }
     return ss::with_gate(
-      _dispatch_gate,
-      [this, b = std::move(b), opts = std::move(opts)]() mutable {
-          if (_correlations.find(_correlation_idx + 1) != _correlations.end()) {
-              _probe.client_correlation_error();
-              throw std::runtime_error(
-                "Invalid transport state. Doubly registered correlation_id");
-          }
-          const uint32_t idx = ++_correlation_idx;
-          auto item = std::make_unique<internal::response_handler>();
-          auto it = item.get();
-          // capture the future _before_ inserting promise in the map
-          // in case there is a concurrent error w/ the connection and it fails
-          // the future before we return from this function
-          auto fut = it->get_future();
-          b.set_correlation_id(idx);
-          auto [_, success] = _correlations.emplace(idx, std::move(item));
-          if (unlikely(!success)) {
-              return ss::make_exception_future<ret_t>(std::logic_error(
-                fmt::format("Tried to reuse correlation id: {}", idx)));
-          }
-          it->with_timeout(opts.timeout, [this, idx] {
-              auto it = _correlations.find(idx);
-              if (likely(it != _correlations.end())) {
-                  vlog(
-                    rpclog.debug, "Request timeout, correlation id: {}", idx);
-                  _probe.request_timeout();
-                  _correlations.erase(it);
-              }
-          });
+             _dispatch_gate,
+             [this, b = std::move(b), opts = std::move(opts), seq]() mutable {
```

Comment:
We should move this to a method tho. it's already a buffer and seems like already type erased.